README.md
[![pypardiso-tests](https://github.com/haasad/PyPardisoProject/actions/workflows/tests.yaml/badge.svg?branch=master)](https://github.com/haasad/PyPardisoProject/actions/workflows/tests.yaml)
# PyPardiso

PyPardiso is a python package to solve large sparse linear systems of equations with the [Intel oneAPI Math Kernel Library PARDISO solver](https://www.intel.com/content/www/us/en/develop/documentation/onemkl-developer-reference-fortran/top/sparse-solver-routines/onemkl-pardiso-parallel-direct-sparse-solver-iface.html), a shared-memory multiprocessing parallel direct sparse solver.

PyPardiso provides the same functionality as SciPy's [scipy.sparse.linalg.spsolve](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.spsolve.html#scipy.sparse.linalg.spsolve) for solving the sparse linear system `Ax=b`. However in many cases it is significantly faster than SciPy's built-in single-threaded SuperLU solver.

PyPardiso is not a python interface to the PARDISO Solver from the [PARDISO 7.2 Solver Project](https://www.pardiso-project.org/) and it also doesn't currently support complex numbers. Check out [JuliaSparse/Pardiso.jl](https://github.com/JuliaSparse/Pardiso.jl/) for these more advanced use cases.

## Installation

PyPardiso runs on Linux, Windows and MacOS. It can be installed with __conda__ or __pip__. It is recommended to install PyPardiso using a virtual environment.

conda-forge | PyPI
:---:|:---:
[![conda-forge version](https://anaconda.org/conda-forge/pypardiso/badges/version.svg)](https://anaconda.org/conda-forge/pypardiso) | [![PyPI version](https://badge.fury.io/py/pypardiso.svg)](https://pypi.org/project/pypardiso/)
`conda install -c conda-forge pypardiso` | `pip install pypardiso`


## Basic usage

How to solve the sparse linear system `Ax=b` for `x`, where `A` is a square, sparse matrix in CSR (or CSC) format and `b` is a vector (or matrix):
```python
In [1]: import pypardiso

In [2]: import numpy as np

In [3]: import scipy.sparse as sp

In [4]: A = sp.rand(10, 10, density=0.5, format='csr')

In [5]: A
Out[5]:
<10x10 sparse matrix of type '<class 'numpy.float64'>'
	with 50 stored elements in Compressed Sparse Row format>

In [6]: b = np.random.rand(10)

In [7]: x = pypardiso.spsolve(A, b)

In [8]: x
Out[8]:
array([ 0.02918389,  0.59629935,  0.33407289, -0.48788966,  3.44508841,
        0.52565687, -0.48420646,  0.22136413, -0.95464127,  0.58297397])
```


CHANGELOG.md
## Changelog

__v0.4.0__

- remove mkl-service dependency (fixes [#24](https://github.com/haasad/PyPardisoProject/issues/24) and [#26](https://github.com/haasad/PyPardisoProject/issues/26))
    - `get_max_threads` and `set_num_threads` methods are removed from the `PyPardisoSolver` class, use `mkl-service` directly if you require them
- change how the shared_library `mkl_rt` is detected to account for differences between conda and pip
    - also removes the psutil dependency

__v0.3.3__

* Release on PyPI and anaconda.org/haasad with github actions (see #19 and #20)

__v0.3.2__

* Change requirements in setup.py to fix [failing conda-forge build](https://github.com/conda-forge/staged-recipes/pull/15106#discussion_r654696146)

__v0.3.1__

- Revert to the old way of detecting the mkl_rt library on osx, since psutil doesn't work (see [#14])

__v0.3.0__

- Changed how pypardiso detects the __mkl_rt__ library to fix a breaking change on windwos with [mkl 2021.2.0](https://anaconda.org/conda-forge/mkl). See [#12](https://github.com/haasad/PyPardisoProject/issues/12) for details.

__v0.2.2__

- CSR-matrix format is forced in `spsolve` and `factorized`. This fixes a serious compatibility issue with [brightway2](https://brightwaylca.org), where a technosphere matrix in CSC-format produces wrong results, due to the bad conditioning of the matrix (see details in issue #7).

__v0.2.1__

- Switched from zero- to one-based indexing for the call to the pardiso library. This brings performance of the factorization phase back to the level of v0.1.0, v0.2.0 is much slower.


tests/utils.py
# coding: utf-8
import numpy as np
import scipy.sparse as sp
from pypardiso.scipy_aliases import pypardiso_solver

ps = pypardiso_solver


def create_test_A_b_small(matrix=False, sort_indices=True):
    """
    --- A ---
    scipy.sparse.csr.csr_matrix, float64
    matrix([[5, 1, 0, 0, 0],
            [0, 6, 2, 0, 0],
            [0, 0, 7, 3, 0],
            [0, 0, 0, 8, 4],
            [0, 0, 0, 0, 9]])
    --- b ---
    np.ndarray, float64
    array([[ 1],
           [ 4],
           [ 7],
           [10],
           [13]])
    or
    array([[ 1,  2,  3],
           [ 4,  5,  6],
           [ 7,  8,  9],
           [10, 11, 12],
           [13, 14, 15]])

    """
    A = sp.spdiags(np.arange(10, dtype=np.float64).reshape(2, 5), [1, 0], 5, 5, format='csr')
    if sort_indices:
        A.sort_indices()
    b = np.arange(1, 16, dtype=np.float64).reshape(5, 3)
    if matrix:
        return A, b
    else:
        return A, b[:, [0]]


def create_test_A_b_rand(n=1000, density=0.05, matrix=False):
    np.random.seed(27)
    A = sp.csr_matrix(sp.rand(n, n, density) + sp.eye(n))
    if matrix:
        b = np.random.rand(n, 5)
    else:
        b = np.random.rand(n, 1)

    return A, b


def basic_solve(A, b):
    x = ps.solve(A, b)
    np.testing.assert_array_almost_equal(A*x, b)


tests/test_complex.py
import numpy as np
import scipy.sparse as sp
from pypardiso import PyPardisoSolver

N = int(1e3)
density = 0.01

# Complex symmetric
A = sp.csr_matrix(sp.rand(N, N, density)+1j*sp.rand(N, N, density)+sp.eye(N))
A = A + A.transpose()
b = np.random.rand(N,1)+1j*np.random.rand(N,1)

ps = PyPardisoSolver(mtype=3)
x = ps.solve(A, b)
np.testing.assert_array_almost_equal(A*x, b)
 
# Complex unsymmetric   
A = sp.csr_matrix(sp.rand(N, N, density)+1j*sp.rand(N, N, density)+sp.eye(N))
b = np.random.rand(N,1)+1j*np.random.rand(N,1)

ps = PyPardisoSolver(mtype=13)
x = ps.solve(A, b)
np.testing.assert_array_almost_equal(A*x, b)

ci/appveyor/install.ps1
# Sample script to install Miniconda under Windows
# Authors: Olivier Grisel, Jonathan Helmus and Kyle Kastner, Robert McGibbon
# License: CC0 1.0 Universal: http://creativecommons.org/publicdomain/zero/1.0/

$MINICONDA_URL = "http://repo.continuum.io/miniconda/"


function DownloadMiniconda ($python_version, $platform_suffix) {
    $webclient = New-Object System.Net.WebClient
    if ($python_version -match "3.5") {
        $filename = "Miniconda3-latest-Windows-" + $platform_suffix + ".exe"
    } else {
        $filename = "Miniconda-latest-Windows-" + $platform_suffix + ".exe"
    }
    $url = $MINICONDA_URL + $filename

    $basedir = $pwd.Path + "\"
    $filepath = $basedir + $filename
    if (Test-Path $filename) {
        Write-Host "Reusing" $filepath
        return $filepath
    }

    # Download and retry up to 3 times in case of network transient errors.
    Write-Host "Downloading" $filename "from" $url
    $retry_attempts = 2
    for($i=0; $i -lt $retry_attempts; $i++){
        try {
            $webclient.DownloadFile($url, $filepath)
            break
        }
        Catch [Exception]{
            Start-Sleep 1
        }
   }
   if (Test-Path $filepath) {
       Write-Host "File saved at" $filepath
   } else {
       # Retry once to get the error message if any at the last try
       $webclient.DownloadFile($url, $filepath)
   }
   return $filepath
}


function InstallMiniconda ($python_version, $architecture, $python_home) {
    Write-Host "Installing Python" $python_version "for" $architecture "bit architecture to" $python_home
    if (Test-Path $python_home) {
        Write-Host $python_home "already exists, skipping."
        return $false
    }
    if ($architecture -match "32") {
        $platform_suffix = "x86"
    } else {
        $platform_suffix = "x86_64"
    }

    $filepath = DownloadMiniconda $python_version $platform_suffix
    Write-Host "Installing" $filepath "to" $python_home
    $install_log = $python_home + ".log"
    $args = "/S /D=$python_home"
    Write-Host $filepath $args
    Start-Process -FilePath $filepath -ArgumentList $args -Wait -Passthru
    if (Test-Path $python_home) {
        Write-Host "Python $python_version ($architecture) installation complete"
    } else {
        Write-Host "Failed to install Python in $python_home"
        Get-Content -Path $install_log
        Exit 1
    }
}


function InstallCondaPackages ($python_home, $spec) {
    $conda_path = $python_home + "\Scripts\conda.exe"
    $args = "install --yes " + $spec
    Write-Host ("conda " + $args)
    Start-Process -FilePath "$conda_path" -ArgumentList $args -Wait -Passthru
}

function UpdateConda ($python_home) {
    $conda_path = $python_home + "\Scripts\conda.exe"
    Write-Host "Updating conda..."
    $args = "update --yes conda"
    Write-Host $conda_path $args
    Start-Process -FilePath "$conda_path" -ArgumentList $args -Wait -Passthru
}


function main () {
    InstallMiniconda $env:PYTHON_VERSION $env:PYTHON_ARCH $env:PYTHON
    UpdateConda $env:PYTHON
    InstallCondaPackages $env:PYTHON "conda-build anaconda-client"
}

main

pypardiso/scipy_aliases.py
# coding: utf-8
import functools
import scipy.sparse as sp
from .pardiso_wrapper import PyPardisoSolver


# pypardsio_solver is used for the 'spsolve' and 'factorized' functions. Python crashes on windows if multiple
# instances of PyPardisoSolver make calls to the Pardiso library
pypardiso_solver = PyPardisoSolver()


def spsolve(A, b, factorize=True, squeeze=True, solver=pypardiso_solver, *args, **kwargs):
    """
    This function mimics scipy.sparse.linalg.spsolve, but uses the Pardiso solver instead of SuperLU/UMFPACK

        solve Ax=b for x

        --- Parameters ---
        A: sparse square CSR or CSC matrix (scipy.sparse.csr.csr_matrix)
        b: numpy ndarray
           right-hand side(s), b.shape[0] needs to be the same as A.shape[0]
        factorize: boolean, default True
                   matrix A is factorized by default, so the factorization can be reused
        squeeze: default True
                 strange quirk of scipy spsolve, which always returns x.squeeze(), this
                 feature in order to keep it compatible with implementations that rely on
                 this behaviour
        solver: instance of PyPardisoSolver, default pypardiso_solver
                you can supply your own instance of PyPardisoSolver, but using several instances
                of PyPardisoSolver in parallel can lead to errors

        --- Returns ---
        x: numpy ndarray
           solution of the system of linear equations, same shape as b (but returnsshape (n,) if b has shape (n,1))

        --- Notes ---
        The computation time increases only minimally if the factorization and the solve phase are carried out
        in two steps, therefore it is factorized by default. Subsequent calls to spsolve with the same matrix A
        will be drastically faster. This makes the "factorized" method obsolete, but it is kept for compatibility.
    """
    if sp.isspmatrix_csc(A):
        A = A.tocsr()  # fixes issue with brightway2 technosphere matrix

    solver._check_A(A)
    if factorize and not solver._is_already_factorized(A):
        solver.factorize(A)

    x = solver.solve(A, b)

    if squeeze:
        return x.squeeze()  # scipy spsolve always returns vectors with shape (n,) indstead of (n,1)
    else:
        return x


def factorized(A, solver=pypardiso_solver, *args, **kwargs):
    """
    This function mimics scipy.sparse.linalg.factorized, but uses the Pardiso solver instead of SuperLU/UMFPACK

        --- Parameters ---
        A: sparse square CSR or CSC matrix (scipy.sparse.csr.csr_matrix)

        --- Returns ---
        solve_b: callable
                 an array b passed to this callable returns the solution to Ax=b

        --- Notes ---
        The returned callable will store a copy of matrix A. This ensures correct results even when the solver
        was used for other tasks in between calls to the returned callable. The factorization is however not
        stored and the first call will take longer.
        Don't use the factorized method for very large matrices, because it needs to keep an additional copy of
        A in memory.
        !!! Use spsolve directly whenever possible !!! Contrary to the scipy implementation there is no performance
        gain in PyPardiso by using factorized instead of spsolve.

    """
    solve_b = functools.partial(spsolve, A.tocsr().copy(), squeeze=False, solver=solver)

    return solve_b


pypardiso/pypardiso.py

# coding: utf-8

import sys
import ctypes
import warnings
import functools
import mkl
import numpy as np
import scipy.sparse as sp
from scipy.sparse import SparseEfficiencyWarning


__all__ = ['PyPardisoSolver', 'spsolve', 'factorized']


class PyPardisoSolver:
    """
    Python interface to the Intel MKL PARDISO library for solving large sparse linear systems of equations Ax=b.
    
    Pardiso documentation: https://software.intel.com/en-us/node/470282
    
    --- Basic usage ---
    matrix type: real (float64) and nonsymetric
    methods: solve, factorize
    
    - use the "solve(A,b)" method to solve Ax=b for x, where A is a sparse CSR matrix and b is a vector or matrix
    - use the "factorize(A)" method first, if you intend to solve the system more than once for different right-hand 
      sides, the factorization will be reused automatically afterwards
    
    
    --- Advanced usage ---
    methods: get_iparm, get_iparms, set_iparm, set_matrix_type, set_phase
    
    - additional options can be accessed by setting the iparms (see Pardiso documentation for description)
    - other matrix types can be chosen with the "set_matrix_type" method. complex matrix types are currently not
      supported
    - the solving phases can be set with the "set_phase" method
    - The out-of-core (OOC) solver either fails or crashes my computer, be careful with iparm[60]
    
    
    --- Statistical info ---
    methods: set_statistical_info_on, set_statistical_info_off
    
    - the Pardiso solver writes statistical info to the C stdout if desired
    - if you use pypardiso from within a jupyter notebook you can turn the statistical info on and capture the output
      real-time by wrapping your call to "solve" with wurlitzer.sys_pipes() (https://github.com/minrk/wurlitzer,
      https://pypi.python.org/pypi/wurlitzer/)
    - wurlitzer dosen't work on windows, info appears in notebook server console window if used from jupyter notebook
    
    
    --- Number of threads ---
    methods: get_max_threads, set_num_threads
    - you can control the number of threads by using the "set_num_threads" method
    
    """
    
    def __init__(self, mtype=11, phase=13):
        if sys.platform == 'darwin':
            libmkl_core = ctypes.CDLL('libmkl_core.dylib')
        elif sys.platform == 'win32':
            libmkl_core = ctypes.CDLL('mkl_core.dll')
        else:
            libmkl_core = ctypes.CDLL('libmkl_core.so')
        
        self._mkl_pardiso = libmkl_core.mkl_pds_lp64_pardiso
        
        self.pt = np.zeros(64, dtype=np.int32)
        self.iparm = np.zeros(64, dtype=np.int32)
        self.perm = np.zeros(0, dtype=np.int32)
        
        self.mtype = mtype
        self.phase = phase
        self.msglvl = False
        
        self.factorized_A = None
        
    
    def factorize(self, A):
        """ 
        Factorize the matrix A, the factorization will automatically be used if the same matrix A is passed to the
        solve method. This will drastically increase the speed of solve, if solve is called more than once for the 
        same matrix A
        
        --- Parameters ---
        A: sparse square CSR matrix (scipy.sparse.csr.csr_matrix)
           other matrix types will be converted to CSR
        """
        self.factorized_A = A
        A = self._check_A(A)
        self.set_phase(12)
        b = np.zeros((A.shape[0],1))
        self._call_pardiso(A, b)    
    
    
    def solve(self, A, b):
        """
        solve Ax=b for x
        
        --- Parameters ---
        A: sparse square CSR matrix (scipy.sparse.csr.csr_matrix)
           other matrix types will be converted to CSR
        b: numpy ndarray
           right-hand side(s), b.shape[0] needs to be the same as A.shape[0]
           
        --- Returns ---
        x: numpy ndarray
           solution of the system of linear equations, same shape as b
        """
        if self.factorized_A is A:
            self.set_phase(33)
        else:
            self.set_phase(13)
        
        A = self._check_A(A)
        b = self._check_b(A, b)
        x = self._call_pardiso(A, b)
        
        return x
    
    
    def _check_A(self, A):
        if A.shape[0] != A.shape[1]:
            raise ValueError('Matrix A needs to be square, but has shape: {}'.format(A.shape))
        
        if not sp.isspmatrix_csr(A):
            warnings.warn('pypardiso requires matrix A to be in CSR format for maximum efficiency', 
                          SparseEfficiencyWarning)
            A = sp.csr_matrix(A)
        
        if A.dtype != np.float64:
            raise TypeError('pypardiso currently only works with float64. Matrix A has dtype: {}'.format(A.dtype))
        
        return A
            
            
    def _check_b(self, A, b):
        if sp.isspmatrix(b):
            warnings.warn('pypardiso requires the right-hand side b to be a dense vector/matrix'
                          ' for maximum efficiency', 
                          SparseEfficiencyWarning)
            b = b.todense()
        
        if b.ndim == 1:
            b = b.reshape(len(b), 1)
        
        if b.shape[0] != A.shape[0]:
            raise ValueError("Dimension mismatch: Matrix A {} and vector/matrix b {}".format(A.shape, b.shape))
            
        if b.dtype != np.float64:
            raise TypeError(('pypardiso currently only works with float64. '
                             'Vector/matrix b has dtype: {}'.format(b.dtype)))
        
        return b
        
    
    def _call_pardiso(self, A, b):
        A_data = A.data
        A_ia = A.indptr + 1
        A_ja = A.indices + 1
        x = np.zeros_like(b)
        pardiso_error = ctypes.c_int32(0)
        
        c_int32_p = ctypes.POINTER(ctypes.c_int32)
        c_float64_p = ctypes.POINTER(ctypes.c_double)

        self._mkl_pardiso(self.pt.ctypes.data_as(c_int32_p), # pt
                          ctypes.byref(ctypes.c_int32(1)), # maxfct
                          ctypes.byref(ctypes.c_int32(1)), # mnum
                          ctypes.byref(ctypes.c_int32(self.mtype)), # mtype -> 11 for real-nonsymetric
                          ctypes.byref(ctypes.c_int32(self.phase)), # phase -> 13 
                          ctypes.byref(ctypes.c_int32(A.shape[0])), #N -> number of equations/size of matrix
                          A_data.ctypes.data_as(c_float64_p), # A -> non-zero entries in matrix
                          A_ia.ctypes.data_as(c_int32_p), # ia -> csr-indptr
                          A_ja.ctypes.data_as(c_int32_p), # ja -> csr-indices
                          self.perm.ctypes.data_as(c_int32_p), # perm -> empty
                          ctypes.byref(ctypes.c_int32(b.shape[1])), # nrhs -> number of right-hand sides
                          self.iparm.ctypes.data_as(c_int32_p), # iparm-array
                          ctypes.byref(ctypes.c_int32(self.msglvl)), # msg-level -> 1: statistical info is printed
                          b.ctypes.data_as(c_float64_p), # b -> right-hand side vector/matrix
                          x.ctypes.data_as(c_float64_p), # x -> output
                          ctypes.byref(pardiso_error)) # pardiso error
        
        if pardiso_error.value != 0:
            raise PyPardisoError(pardiso_error.value)
        else:
            return x
        
        
    def get_iparms(self):
        """Returns a dictionary of iparms"""
        return dict(enumerate(self.iparm, 1))
    
    
    def get_iparm(self, i):
        """Returns the i-th iparm (1-based indexing)"""
        return self.iparm[i-1]
    
    
    def set_iparm(self, i, value):
        """set the i-th iparm to 'value' (1-based indexing)"""
        if not i in {1,2,4,5,6,8,10,11,12,13,18,19,21,24,25,27,28,31,34,35,36,37,56,60}:
            warnings.warn('{} is no input iparm. See the Pardiso documentation.'.format(value), UserWarning)
        self.iparm[i-1] = value
        
        
    def set_matrix_type(self, mtype):
        """Set the matrix type (see Pardiso documentation)"""
        self.mtype = mtype
        
        
    def get_max_threads(self):
        """Returns the maximum number of threads the solver will use"""
        return mkl.get_max_threads()
    
    
    def set_num_threads(self, num_threads):
        """Set the number of threads the solver should use (only a hint, not guaranteed that 
        the solver uses this amount)"""
        mkl.set_num_threads(num_threads)
        
    
    def set_statistical_info_on(self):
        """Display statistical info (appears in notebook server console window if pypardiso is 
        used from jupyter notebook, use wurlitzer to redirect info to the notebook)"""
        self.msglvl = 1
        
        
    def set_statistical_info_off(self):
        """Turns statistical info off"""
        self.msglvl = 0
        
    
    def set_phase(self, phase):
        """Set the phase(s) for the solver. See the Pardiso documentation for details."""
        self.phase = phase


# pypardsio_solver is used for the 'spsolve' and 'factorized' functions. Python crashes on windows if multiple 
# instances of PyPardisoSolver make calls to the Pardiso library
pypardiso_solver = PyPardisoSolver()


def spsolve(A, b, factorize=True):
    """
    This function mimics scipy.sparse.linalg.spsolve, but uses the Pardiso solver instead of SuperLU/UMFPACK
    
        solve Ax=b for x
        
        --- Parameters ---
        A: sparse square CSR matrix (scipy.sparse.csr.csr_matrix)
           other matrix types will be converted to CSR
        b: numpy ndarray
           right-hand side(s), b.shape[0] needs to be the same as A.shape[0]
        factorize: boolean, default True
                   matrix A is factorized by default, so the factorization can be reused
           
        --- Returns ---
        x: numpy ndarray
           solution of the system of linear equations, same shape as b
           
        --- Notes ---
        the computation time increases only minimally if the factorization and the solve phase are carried out 
        in two steps, therefore it is factorized by default. Subsequent calls to spsolve with the same matrix A 
        will be drastically faster
    """
    if factorize and not pypardiso_solver.factorized_A is A:
        pypardiso_solver.factorize(A)
    x = pypardiso_solver.solve(A, b)
    return x


def factorized(A):
    """
    This function mimics scipy.sparse.linalg.factorized, but uses the Pardiso solver instead of SuperLU/UMFPACK
    
        --- Parameters ---
        A: sparse square CSR matrix (scipy.sparse.csr.csr_matrix)
           other matrix types will be converted to CSR
           
        --- Returns ---
        solve_b: callable 
        		 a vector/matrix b passed to this callable returns the solution to Ax=b
    """
    pypardiso_solver.factorize(A)
    solve_b = functools.partial(pypardiso_solver.solve, A)
    return solve_b


class PyPardisoError(Exception):
    
    def __init__(self, value):
        self.value = value
    
    def __str__(self):
        return ('The Pardiso solver failed with error code {}. '
                'See Pardiso documentation for details.'.format(self.value))

